File: 0687-longest-univalue-path/0687-longest-univalue-path.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = 0;
    public int longestUnivaluePath(TreeNode root) {
        if(root == null) return 0;
        dfs(root, root.val);
        return ans;
    }
    private int dfs(TreeNode node, int val) {
        if(node == null) return 0;
        int lv = dfs(node.left, node.val);
        int rv = dfs(node.right, node.val);
        ans = Math.max(lv+rv, ans);
        if(node.val == val) return Math.max(lv, rv) + 1;
        return 0;
    }
}

File: 0687-longest-univalue-path/0687-longest-univalue-path.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function longestUnivaluePath(root: TreeNode | null): number {
    var ans = 0;
    if(root == null) return 0;
    var dfs = function(node, val) {
        if(node == null) return 0;
        var lv = dfs(node.left, node.val);
        var rv = dfs(node.right, node.val);
        ans = Math.max(lv+rv, ans);
        if(node.val == val) return Math.max(lv, rv) + 1;
        return 0;
    }
    dfs(root, root.val)
    return ans
};